adapters/base/src/main/java/org/apache/cassandra/sidecar/adapters/base/TokenRangeReplicas.java

```diff
@@ -0,0 +1,472 @@
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o)
+        {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass())
+        {
+            return false;
+        }
+
+        TokenRangeReplicas that = (TokenRangeReplicas) o;
+
+        return Objects.equals(start, that.start)
+               && Objects.equals(end, that.end)
+               && partitioner == that.partitioner;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(start, end, partitioner);
+    }
+
+    private void validateRangesForComparison(@NotNull TokenRangeReplicas other)
+    {
+        if (this.partitioner != other.partitioner)
+            throw new IllegalStateException("Token ranges being compared do not have the same partitioner");
+    }
+
+    protected boolean contains(TokenRangeReplicas other)
+    {
+        validateRangesForComparison(other);
+        return (other.start.compareTo(this.start) >= 0 && other.end.compareTo(this.end) <= 0);
+    }
+
+    /**
+     * For subset ranges, this is used to determine if a range is larger than the other by comparing start-end lengths
+     * If both ranges end at the min, we compare starting points to determine the result.
+     * When the left range is the only one ending at min, it is always the larger one since all subsequent ranges
+     * in the sorted range list have to be smaller.
+     * <p>
+     * This method assumes that the ranges are normalized and unwrapped, i.e.
+     * 'this' comes before 'other' AND there's no wrapping around the min token
+     *
+     * @param other the next range in the range list to compare
+     * @return true if "this" range is larger than the other
+     */
+    protected boolean isLarger(TokenRangeReplicas other)
+    {
+        validateRangesForComparison(other);
+        return this.end.subtract(this.start).compareTo(other.end.subtract(other.start)) > 0;
+    }
+
+    /**
+     * Determines intersection if the next range starts before the current range ends. This method assumes that
+     * the provided ranges are sorted and unwrapped.
+     * When the current range goes all the way to the end, we determine intersection if the next range starts
+     * after the current since all subsequent ranges have to be subsets.
+     *
+     * @param other the range we are currently processing to check if "this" intersects it
+     * @return true if "this" range intersects the other
+     */
+    protected boolean intersects(TokenRangeReplicas other)
+    {
+        if (this.compareTo(other) > 0)
+            throw new IllegalStateException(
+            String.format("Token ranges - (this:%s other:%s) are not ordered", this, other));
+
+        return this.end.compareTo(other.start) > 0 && this.start.compareTo(other.end) < 0; // Start exclusive (DONE)
+    }
```

Comment:
Rather then throwing if this is < other, how about just reversing the comparison in the return statement:

```suggestion
    protected boolean intersects(TokenRangeReplicas other)
    {
        boolean inOrder = this.compareTo(other) <= 0;
        TokenRangeReplicas first = inOrder ? this : other;
        TokenRangeReplicas last = inOrder ? other : this;

        return first.end.compareTo(last.start) > 0 && first.start.compareTo(last.end) < 0; // Start exclusive (DONE)
    }
```

Reply:
Apart from here there are other operations that rely on the ranges being processed to be sorted and breaks if the prerequisite is not met. Although that should "never" happen since we are explicitly sorting and unwrapping upfront. 

Like you mentioned earlier, I'm leaning towards performing the assertions on the assumptions up-front, if for any reason the ranges are not what we expect them to be.

Reply:
Another option would be to perform the assertion within `processIntersectingRanges` instead and do what you're suggesting in the `intersects()` method. That way, we only fail in the event there are intersections (and the ranges are out of order).